target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,651 @@
+	clk->idx = idx;
+
+	parent_data.fw_name = rtcl_clk_info[idx].parent_name;
+
+	hw_init.name = rtcl_clk_info[idx].name;
+	hw_init.ops = &rtcl_clk_ops;
+	hw_init.parent_data = &parent_data;
+	hw_init.num_parents = 1;
+
+	ret = of_clk_hw_register(rtcl_ccu->np, &clk->hw);
+	if (ret)
+		return ret;
+
+	clk_hw_register_clkdev(&clk->hw, rtcl_clk_info[idx].name, NULL);
+
+	return 0;
+}
+
+static struct clk_hw *rtcl_get_clkhw(struct of_phandle_args *clkspec, void *priv)
+{
+	unsigned int idx = clkspec->args[0];
+
+	if (idx >= CLK_COUNT) {
+		pr_err("%s: Invalid index %u\n", __func__, idx);
+		return ERR_PTR(-EINVAL);
+	}
+
+	return &rtcl_ccu->clks[idx].hw;
+}
+
+static int rtcl_ccu_register_clocks(void)
+{
+	int idx, ret;
+
+	for (idx = 0; idx < CLK_COUNT; ++idx) {
+		ret = rtcl_register_clkhw(idx);
+		if (ret) {
+			pr_err("%s: Couldn't register %s clock\n",
+				__func__, rtcl_clk_info[idx].display_name);
+			goto err_hw_unregister;
+		}
+	}
+
+	ret = of_clk_add_hw_provider(rtcl_ccu->np, rtcl_get_clkhw, rtcl_ccu);
+	if (ret) {
+		pr_err("%s: Couldn't register clock provider of %s\n",
+			__func__, of_node_full_name(rtcl_ccu->np));
+		goto err_hw_unregister;
+	}
+
+	return 0;
+
+err_hw_unregister:
+	for (--idx; idx >= 0; --idx)
+		clk_hw_unregister(&rtcl_ccu->clks[idx].hw);
+
+	return ret;
+}
+
+void rtcl_ccu_log(void)
+{
+	int idx, rate;
+	struct clk *clk;
+	char msg[255], msgl[255] = "rtl83xx-clk: initialized";
+
+	for (idx = 0; idx < CLK_COUNT; idx++) {
+		clk = clk_get(NULL, rtcl_clk_info[idx].name);
+		rate = clk_get_rate(clk);
+		clk_put(clk);
+
+		sprintf(msg, ", %s %d MHz", rtcl_clk_info[idx].display_name, rate / 1000000);
+		strcat(msgl, msg);
+	}
+	pr_info("%s\n", msgl);
+}
+
+int rtcl_init_sram(void)
+{
+	struct gen_pool *sram_pool;
+	phys_addr_t sram_pbase;
+	unsigned long sram_base;
+	struct device_node *node;
+	struct platform_device *pdev = NULL;
+	void *dram_start;
+	int dram_size;
+	char msge[80] = ", rate setting disabled.\n";
+	char msgi[80] = "rate setting enabled\n";
+
+	switch (rtcl_ccu->soc) {
+	case SOC_RTL838X:
+		dram_start = &rtcl_838x_dram_start;
+		dram_size = rtcl_838x_dram_size;
+		break;
+	case SOC_RTL839X:
+		dram_start = &rtcl_839x_dram_start;
+		dram_size = rtcl_839x_dram_size;
+		break;
+	default:
+		return -ENXIO;
+	}
+
+	if (!dram_size) {
+		dev_info(&rtcl_ccu->pdev->dev, "%s", msgi);
+		rtcl_ccu->sramok = true;
+		return 0;
+	}
+
+	for_each_compatible_node(node, NULL, "mmio-sram") {
+		pdev = of_find_device_by_node(node);
+		if (pdev) {
+			of_node_put(node);
+			break;
+		}
+	}
+	if (!pdev) {
+		dev_warn(&rtcl_ccu->pdev->dev, "no SRAM device found%s", msge);
+		return -ENXIO;
+	}
+
+	sram_pool = gen_pool_get(&pdev->dev, NULL);
+	if (!sram_pool) {
+		dev_warn(&rtcl_ccu->pdev->dev, "SRAM pool unavailable%s", msge);
+		goto err_put_device;
+	}
+
+	sram_base = gen_pool_alloc(sram_pool, dram_size);
+	if (!sram_base) {
+		dev_warn(&rtcl_ccu->pdev->dev, "unable to allocate SRAM%s", msge);
+		goto err_put_device;
+	}
```

Comment:
If I got this right, reclocking from SRAM is just disabled if the required code is absent (makes sense). If reclocking is provided, but the SRAM node is missing, you fail driver probing however. AFAICT, SRAM doesn't appear to be used for reclocking on RTL838x, so are you making driver probing on RTL838x depend on something that isn't required? Not that I'm against adding the SRAM node to the RTL838x DTSI; that would still serve as good documentation.

Reply:
838x will need SRAM in the forthcoming PR version too.